pkg/controller/persistentvolumelabel/pvlcontroller.go

```diff
@@ -0,0 +1,421 @@
+	if pvlc.shouldLabel(volume) {
+		if volume.Spec.AWSElasticBlockStore != nil {
+			labels, err := pvlc.findAWSEBSLabels(volume)
+			if err != nil {
+				return fmt.Errorf("error querying AWS EBS volume %s: %v", volume.Spec.AWSElasticBlockStore.VolumeID, err)
+			}
+			volumeLabels = labels
+		}
+		if volume.Spec.GCEPersistentDisk != nil {
+			labels, err := pvlc.findGCEPDLabels(volume)
+			if err != nil {
+				return fmt.Errorf("error querying GCE PD volume %s: %v", volume.Spec.GCEPersistentDisk.PDName, err)
+			}
+			volumeLabels = labels
+		}
+
+		if len(volumeLabels) != 0 {
+			if volume.Labels == nil {
+				volume.Labels = make(map[string]string)
+			}
+			for k, v := range volumeLabels {
+				// We (silently) replace labels if they are provided.
+				// This should be OK because they are in the kubernetes.io namespace
+				// i.e. we own them
+				volume.Labels[k] = v
+			}
+		}
+		return pvlc.updateVolume(volume)
+	}
+
+	return nil
+}
+
+func (pvlc *PersistentVolumeLabelController) findAWSEBSLabels(volume *v1.PersistentVolume) (map[string]string, error) {
+	// Ignore any volumes that are being provisioned
+	if volume.Spec.AWSElasticBlockStore.VolumeID == vol.ProvisionedVolumeName {
+		return nil, nil
+	}
+	ebsVolumes, err := pvlc.getEBSVolumes()
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: GetVolumeLabels is actually a method on the Volumes interface
+	// If that gets standardized we can refactor to reduce code duplication
+	spec := aws.KubernetesVolumeID(volume.Spec.AWSElasticBlockStore.VolumeID)
+	labels, err := ebsVolumes.GetVolumeLabels(spec)
+	if err != nil {
+		return nil, err
+	}
+
+	return labels, nil
+}
+
+// getEBSVolumes returns the AWS Volumes interface for ebs
+func (pvlc *PersistentVolumeLabelController) getEBSVolumes() (aws.Volumes, error) {
+	pvlc.mutex.Lock()
+	defer pvlc.mutex.Unlock()
+
+	if pvlc.ebsVolumes == nil {
+		awsCloudProvider, ok := pvlc.cloud.(*aws.Cloud)
+		if !ok {
+			// GetCloudProvider has gone very wrong
+			return nil, fmt.Errorf("error retrieving AWS cloud provider")
+		}
+		pvlc.ebsVolumes = awsCloudProvider
+	}
+	return pvlc.ebsVolumes, nil
+}
+
+func (pvlc *PersistentVolumeLabelController) findGCEPDLabels(volume *v1.PersistentVolume) (map[string]string, error) {
+	// Ignore any volumes that are being provisioned
+	if volume.Spec.GCEPersistentDisk.PDName == vol.ProvisionedVolumeName {
+		return nil, nil
+	}
+
+	provider, err := pvlc.getGCECloudProvider()
+	if err != nil {
+		return nil, err
+	}
+
+	// If the zone is already labeled, honor the hint
+	zone := volume.Labels[kubeletapis.LabelZoneFailureDomain]
+
+	labels, err := provider.GetAutoLabelsForPD(volume.Spec.GCEPersistentDisk.PDName, zone)
+	if err != nil {
+		return nil, err
+	}
+
+	return labels, nil
+}
+
+// getGCECloudProvider returns the GCE cloud provider, for use for querying volume labels
+func (pvlc *PersistentVolumeLabelController) getGCECloudProvider() (*gce.GCECloud, error) {
+	pvlc.mutex.Lock()
```

Comment:
This is the same impl as is in the admission plugin.  The cloud provider specific code will be going away in a future revision.